sdk/core/src/main/java/com/google/cloud/trace/AbstractTraceContextHandler.java

```diff
@@ -35,31 +36,62 @@
    * the stack and cannot be removed.
    */
   public AbstractTraceContextHandler(TraceContext rootContext) {
-    this.contextStack = new ArrayDeque<TraceContext>(Collections.singleton(rootContext));
+    this.contextStack = new ArrayList<TraceContext>();
+    replace(rootContext);
   }
 
   @Override
   public TraceContext current() {
-    return contextStack.peekFirst();
+    return contextStack.get(contextStack.size() - 1);
   }
 
   @Override
   public void push(TraceContext context) {
-    contextStack.addFirst(context);
+    contextStack.add(context);
     doPush(context);
   }
 
   @Override
   public TraceContext pop() {
-    if (contextStack.size() > 1) {
-      TraceContext context = contextStack.removeFirst();
+    int indexToRemove = contextStack.size() - 1;
+    if (indexToRemove > rootIndex) {
+      TraceContext context = contextStack.remove(indexToRemove);
       doPop(current());
       return context;
     }
     logger.warning("Too many calls to AbstractTraceContextHandler.pop().");
     return null;
   }
 
+  @Override
+  public TraceContextHandlerState replace() {
+    int previousRoot = rootIndex;
+    // The last item is the new root.
+    this.rootIndex = contextStack.size() - 1;
```

Comment:
In this case, the new root is the next item to be added, so just contextStack.size().